### PR TITLE
Publish redux

### PR DESCRIPTION
**Pull request by isuftin**

### CODE REVIEW

File: coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java [hunk 1]
```diff
@@ -154,39 +149,124 @@ public Response updateChildren(@Context HttpServletRequest request, @PathParam("
...
+
+		return response;
+	}
+
+	/**
+	 * Updates one or more items for children and displayed children
+	 *
+	 * @return whether items were updated in the database or not
+	 */
+	boolean updateItemChildren(Map<String, JsonObject> items) {
```
Comment (jiwalker-usgs) [inline coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java]: no visibility on this?

File: coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java [hunk 2]
```diff
@@ -154,39 +149,124 @@ public Response updateChildren(@Context HttpServletRequest request, @PathParam("
...
 					}
 					parentItem.setChildren(children);
-					itemList.add(parentItem);
-					
-					thumbMan.updateDirtyBits(parentItem.getId());
+
+					// Update the item's displayedChildren
+					if (updateData.has("displayedChildren")) {
+						Iterator<JsonElement> displayedIterator = updateData.get("displayedChildren").getAsJsonArray().iterator();
+						List<String> displayedChildren = new ArrayList<>();
```
Comment (jiwalker-usgs) [inline coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java]: This is my fault, but probably confusing to mix Array and Linked lists in same method.

File: coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java [hunk 3]
```diff
@@ -154,39 +149,124 @@ public Response updateChildren(@Context HttpServletRequest request, @PathParam("
...
+		Response response;
+		if (StringUtils.isNotBlank(content)) {
+			try {
+				JsonObject jsonObj = (JsonObject) new JsonParser().parse(content);
+				
+				Set<Entry<String, JsonElement>> entrySet = jsonObj.entrySet();
+				int itemCount = entrySet.size();
+				if (itemCount > 0) {
+					// Create the map that the updating function expects 
+					Map<String, JsonObject> itemMap = new HashMap<>(itemCount);
```
Comment (jiwalker-usgs) [inline coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java]: It's probably not a great idea to set the initial capacity to the number of things you are going to add.  If you are trying to optimize the map set the initial size to itemCount*1.35 so you don't trigger the resize.

File: coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java [hunk 4]
```diff
@@ -154,39 +149,124 @@ public Response updateChildren(@Context HttpServletRequest request, @PathParam("
...
+					for (Entry<String, JsonElement> entry : entrySet) {
+						itemMap.put(entry.getKey(), (JsonObject) entry.getValue());
```
Comment (jiwalker-usgs) [inline coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java]: Rather than cast it here, why not use the getAsJsonObject convenience method?

File: coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java (hunk 3 above)
Comment (isuftin) [inline coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java]: Good point. Do you know if the resize happens when init capacity is reached or when it's like 1 below init capacity or whatever?

File: coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java (hunk 1 above)
Comment (isuftin) [inline coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java]: could've left it private but left it default. When I initially wrote it, I thought it might have been testable but as I refactored it, it isn't testable due to db access methods. Will change to priv.

File: coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java (hunk 4 above)
Comment (isuftin) [inline coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java]: Didn't know that was a thing. Can do that.

File: coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java [hunk 5]
```diff
@@ -154,39 +149,124 @@ public Response updateChildren(@Context HttpServletRequest request, @PathParam("
...
-		JsonElement parsed = parser.parse(content);
-		if (parsed instanceof JsonObject) {
-			JsonObject jsonObj = (JsonObject) parsed;
-
-			try (ItemManager manager = new ItemManager(); ThumbnailManager thumbMan = new ThumbnailManager()) {
-				List<Item> itemList = new LinkedList<>();
-				for (Entry<String, JsonElement> entry : jsonObj.entrySet()) {
-					Item parentItem = manager.load(entry.getKey());
-					List<Item> children = new LinkedList<>();
-					Iterator<JsonElement> iterator = ((JsonArray) entry.getValue()).iterator();
```
Comment (jiwalker-usgs) [inline coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java]: Do we not have JsonArrays anymore?

File: coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java (hunk 3 above)
Comment (jiwalker-usgs) [inline coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java]: It happens when it passes the loadFactor, which is .75 by default `1 / 0.75 = 1.333` I rounded up.

Comment (isuftin) [inline coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java]: Ah, then screw the optimization step here. Will remove.

**Comment by jiwalker-usgs** [comment]
Looks good, my last question relates to whether or not it works mostly.  It is hard to follow whether anything important was removed or not.

**Comment by isuftin** [comment]
It is functioning properly. We've tested on one and multiple items and all updated come through correctly.

File: coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java (hunk 4 above)
Comment (isuftin) [inline coastal-hazards-portal/src/main/java/gov/usgs/cida/coastalhazards/rest/data/TreeResource.java]: Apparently a JsonElement object does not expose getAsJsonObject

**Comment by jiwalker-usgs** [comment]
https://google-gson.googlecode.com/svn/trunk/gson/docs/javadocs/com/google/gson/JsonElement.html#getAsJsonObject()
